src/Features/LanguageServer/ProtocolUnitTests/CodeActions/CodeActionsTests.cs

```diff
@@ -75,6 +78,65 @@ void M()
             AssertJsonEquals(expected, introduceConstant);
         }
 
+        [WpfFact]
+        public async Task TestCodeActionsCacheAsync()
+        {
+            var markup =
+@"class A
+{
+    void M()
+    {
+        {|caret:|}int i = 1;
+    }
+}";
+            using var workspace = CreateTestWorkspace(markup, out var locations);
+            var cache = GetCodeActionsCache(workspace);
+
+            var caretLocation = locations["caret"].Single();
+            await RunGetCodeActionsAsync(workspace.CurrentSolution, caretLocation);
+
+            var document = GetDocument(workspace, CreateTextDocumentIdentifier(caretLocation.Uri));
+            var cacheResults = await cache.GetCacheAsync(document, caretLocation.Range, CancellationToken.None);
+            Assert.NotNull(cacheResults);
+            Assert.True(cache.GetNumCacheItems() == 1);
+
+            // Invoking code actions on the same unmodified document and range should use the existing cached item
+            // instead of generating a new cached item.
+            await RunGetCodeActionsAsync(workspace.CurrentSolution, caretLocation);
+            cacheResults = await cache.GetCacheAsync(document, caretLocation.Range, CancellationToken.None);
+            Assert.True(cache.GetNumCacheItems() == 1);
+
+            // Invoking code actions on a different range should generate a new cached item.
+            caretLocation.Range = new LSP.Range
+            {
+                Start = new LSP.Position() { Line = 0, Character = 0 },
+                End = new LSP.Position() { Line = 0, Character = 0 }
+            };
+
+            await RunGetCodeActionsAsync(workspace.CurrentSolution, caretLocation);
+            Assert.True(cache.GetNumCacheItems() == 2);
+
+            // Changing the document should generate a new cached item.
+            var currentDocText = await document.GetTextAsync();
+            var changedSourceText = currentDocText.WithChanges(new TextChange(new TextSpan(0, 0), "class D { } \n"));
+            var docId = ((TestWorkspace)workspace).Documents.First().Id;
+            ((TestWorkspace)workspace).ChangeDocument(docId, changedSourceText);
+            UpdateSolutionProvider((TestWorkspace)workspace, workspace.CurrentSolution);
+
+            await RunGetCodeActionsAsync(workspace.CurrentSolution, caretLocation);
+            Assert.True(cache.GetNumCacheItems() == 3);
+
+            // The current cache size is 3. Adding a 4th item to the cache should still keep the cache size the same.
+            caretLocation.Range = new LSP.Range
+            {
+                Start = new LSP.Position() { Line = 0, Character = 0 },
+                End = new LSP.Position() { Line = 0, Character = 1 }
+            };
+
+            await RunGetCodeActionsAsync(workspace.CurrentSolution, caretLocation);
+            Assert.True(cache.GetNumCacheItems() == 3);
```

Comment:
Assert that the documents you expect are in the cache? The count is essentially pinned at 3 now, but the tests won't prevent a bug being introduced in future where the wrong things are cached.

This would also be useful to validate the the oldest item is being evicted.

Probably need to add a TestAccessor for all of that, but you could then remove GetNumCacheItems